### PR TITLE
EnergiBridge Integration

### PR DESCRIPTION
**Pull request by mhkarsten**
This PR implements a self-measurement feature for RunnerConfig, such that a user can simply specify self-measure=True, and receive energy measurements in their results.

- Currently only supports Linux (could also support other oses if needed)
- Requires EnergiBridge to already be installed

+ Also implements customizable install location for EnergiBridge, as its user installed it might be in different locations
+ Added config validation for self-measure
+ Added an example of how to enable self-measurement.

### CODE REVIEW

File: experiment-runner/ConfigValidator/Config/Validation/ConfigValidator.py [hunk 1]
```diff
@@ -18,7 +21,47 @@ def __check_expression(name, value, expected, expression):
...
+            or not os.access(eb_path, os.X_OK):
+
+            ConfigValidator.error_found = True
+            ConfigValidator \
+            .config_values_or_exception_dict["EnergiBridge"] = "EnergiBridge executable was not present or valid"
+
+        # Test run to see if energibridge works
+        try:
+            eb_args = [eb_path, "--summary", "-o", "/dev/null", "--", "sleep", "0.5"]
+            p = subprocess.Popen(eb_args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
```
Comment (Raiduy) [inline experiment-runner/ConfigValidator/Config/Validation/ConfigValidator.py:43]: Might need `sudo` privileges

File: experiment-runner/ConfigValidator/Config/Validation/ConfigValidator.py [hunk 2]
```diff
@@ -27,6 +70,13 @@ def validate_config(config: RunnerConfig):
         if '~' in str(config.experiment_path):
             config.experiment_path = config.experiment_path.expanduser()
         
+        # Set defaults to support configs without the self_measure parameter
+        if not hasattr(config, "self_measure"):
+            config.self_measure = False
+
+        if config.self_measure and not hasattr(config, "self_measure_bin"):
+            config.self_measure_bin = "/usr/local/bin/energibridge" # This is spesific to linux, might work for osx as well
```
Comment (Raiduy) [inline experiment-runner/ConfigValidator/Config/Validation/ConfigValidator.py:78]: Apparently the way to make it runnable from anywhere on mac is to `export PATH` (on the newer `zsh` terminals)

File: experiment-runner/ExperimentOrchestrator/Experiment/Run/RunController.py [hunk 3]
```diff
@@ -6,8 +8,45 @@
 from ProgressManager.Output.OutputProcedure import OutputProcedure as output
 
 class RunController(IRunController):
+    # Start EnergiBridge measurements
+    def start_eb(self):
+        if not self.config.self_measure:
+            return
+
+        eb_args = [self.config.self_measure_bin, "--summary",
```
Comment (Raiduy) [inline experiment-runner/ExperimentOrchestrator/Experiment/Run/RunController.py:16]: `sudo` privileges might be needed

File: experiment-runner/ExperimentOrchestrator/Experiment/Run/RunController.py [hunk 4]
```diff
@@ -6,8 +8,45 @@
...
+                   "-o", "/dev/null", "--", "sleep", "1000000"]
+
+        try:
+            self.eb_proc = subprocess.Popen(eb_args, stdout=subprocess.PIPE,
```
Comment (Raiduy) [inline experiment-runner/ExperimentOrchestrator/Experiment/Run/RunController.py:20]: Maybe also add energibridge log file for output

**Review by Raiduy — COMMENTED** [review]
Essentially the same review as on the PICOLOG branch :)

File: experiment-runner/ConfigValidator/Config/Validation/ConfigValidator.py (hunk 1 above)
Comment (mhkarsten) [reply experiment-runner/ConfigValidator/Config/Validation/ConfigValidator.py:43]: So sudo privileges are not needed if you set EnergiBridge up correctly. They instruct you to add the binary to a group that gets access to the features it needs. That being intel RAPL. A model-specific register that allows energy metrics to be polled.

File: experiment-runner/ConfigValidator/Config/Validation/ConfigValidator.py (hunk 2 above)
Comment (mhkarsten) [reply experiment-runner/ConfigValidator/Config/Validation/ConfigValidator.py:78]: I assumed that mac had a PATH env, I wasn't sure if it contained /usr/local/bin (where users on linux put their non-system installed binaries). Sadly I don't have one to test it on :(

File: experiment-runner/ConfigValidator/Config/Validation/ConfigValidator.py (hunk 1 above)
Comment (Raiduy) [reply experiment-runner/ConfigValidator/Config/Validation/ConfigValidator.py:43]: Yep yep, you're right

**Comment by mhkarsten** [comment]
Closing this request, its implementation was moved to the Pico CM3 PR.

**Pull request closed without merging**